File: memory/shift.py

import base
import parser


def random_shift(machine, nxt, rand, cost):
    bits = machine.addr_bits - machine.word_bits - 1
    shift = rand.randint(-bits, bits)
    return Shift(base.Join(), nxt, shift)


class Shift(base.Transform):

    def __init__(self, bank, mem, shift):
        base.Transform.__init__(self, bank, mem)
        self.shift = shift

    def __str__(self):
        result = "(shift "
        result += "(value " + str(self.shift) + ")"
        result += '(bank ' + str(self.bank) + ')'
        result += '(memory ' + str(self.mem) + ')'
        result += ")"
        return result

    def generate(self, gen, mach):
        self.generate_transform("shift", self.shift, -self.shift, gen, mach)

    def is_empty(self):
        return self.shift == 0

    def permute(self, rand, max_cost):
        bits = self.machine.addr_bits - self.machine.word_bits - 1
        self.shift = rand.randint(-bits, bits)
        return True

    def push_transform(self, index, rand):
        if index == 0:
            rand.push_transform(lambda a: self._rotate(a, self.shift))
        else:
            rand.push_transform(lambda a: a)

    def pop_transform(self, rand):
        rand.pop_transform()

    def get_transform_path_length(self):
        return 0

    def _rotate(self, value, count):
        bits = self.machine.addr_bits - self.machine.word_bits
        mask = (1 << bits) - 1
        shift_part = value >> self.machine.word_bits
        word_part = value & self.machine.word_mask
        if count >= 0:
            count2 = bits - count
            shifted = ((shift_part << count) & mask) | (shift_part >> count2)
        else:
            count2 = bits + count
            shifted = ((shift_part << count2) & mask) | (shift_part >> -count)
        return (shifted << self.machine.word_bits) | word_part

    def process(self, start, write, addr, size):
        addr = self._rotate(addr, self.shift)
        return self.bank.process(start, write, addr, size)

    def forward(self, index, start, write, addr, size):
        addr = self._rotate(addr, -self.shift)
        return self.mem.process(start, write, addr, size)


def _create_shift(lexer, args):
    value = parser.get_argument(lexer, args, 'value', 0)
    mem = parser.get_argument(lexer, args, 'memory')
    bank = parser.get_argument(lexer, args, 'bank')
    return Shift(bank, mem, value)
base.constructors['shift'] = _create_shift

File: fastsim/shift.ml
open Machine
open Base_memory

class shift =
    object (self)
        inherit transform as super

        val mutable offset : int = 0

        method set name value =
            match name with
            | "value" -> offset <- int_of_string value
            | _ -> super#set name value

        method private rotate value count =
            let word_size = self#word_size in
            let word_bits = (log2 word_size) - 1 in
            let word_mask = word_size - 1 in
            let bits = mach.addr_bits - word_bits in
            let count = count mod bits in
            let mask = (1 lsl bits) - 1 in
            let addr = value lsr word_bits in
            let word = value land word_mask in
            let shifted =
                if count >= 0 then
                    let right = bits - count in
                    ((addr lsl count) land mask) lor (addr lsr right)
                else
                    let left = bits + count in
                    ((addr lsl left) land mask) lor (addr lsr -count)
            in (shifted lsl word_bits) lor word

        method private process base start write addr size =
            let addr = self#rotate addr offset in
            self#bank#send_request base start write addr size

        method forward base index start write addr size =
            let addr = self#rotate addr (-offset) in
            self#next#send_request base start write addr size

    end
